### PR TITLE
Roomserver perform leave

### PR DESCRIPTION
**Pull request by neilalexander**
Moves the leave logic into the new perform workflow between the roomserver and the federationsender, and out of the client API.

### CODE REVIEW

File: federationsender/internal/perform.go [hunk 1]
```diff
@@ -153,5 +153,79 @@ func (r *FederationSenderInternalAPI) PerformLeave(
...
+		// request.
+		respMakeLeave, err := r.federation.MakeLeave(
+			ctx,
+			serverName,
+			request.RoomID,
+			request.UserID,
+		)
+		if err != nil {
+			// TODO: Check if the user was not allowed to leave the room.
+			return err
```
Comment (Kegsay) [inline federationsender/internal/perform.go:172]: Or try the next server? Justify why returning is the right thing here.

File: roomserver/storage/sqlite3/event_state_keys_table.go [hunk 2]
```diff
@@ -136,7 +136,7 @@ func (s *eventStateKeyStatements) bulkSelectEventStateKey(
 	for k, v := range eventStateKeyNIDs {
 		iEventStateKeyNIDs[k] = v
 	}
-	selectOrig := strings.Replace(bulkSelectEventStateKeyNIDSQL, "($1)", common.QueryVariadic(len(eventStateKeyNIDs)), 1)
+	selectOrig := strings.Replace(bulkSelectEventStateKeySQL, "($1)", common.QueryVariadic(len(eventStateKeyNIDs)), 1)
```
Comment (Kegsay) [inline roomserver/storage/sqlite3/event_state_keys_table.go:139]: Please think of your future self and rename the constant! `bulkSelectEventStateKeyNIDSQL` should be `bulkSelectEventStateKeySQL` and vice versa.

File: roomserver/internal/perform_leave.go [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+
+	// Let's see if we have an event active for the user in the room. If
+	// we do then it will contain a server name that we can direct the
+	// send_leave to.
+	senderUserNIDs, err := r.DB.GetInvitesForUser(ctx, roomNID, targetUserNID)
+	if err != nil {
+		return "", fmt.Errorf("r.DB.GetInvitesForUser: %w", err)
+	}
+	if len(senderUserNIDs) == 0 {
+		return "", fmt.Errorf("no senderUserNIDs")
```
Comment (Kegsay) [inline roomserver/internal/perform_leave.go:190]: Doesn't this just mean `isInvitePending` == `false`?

File: roomserver/storage/sqlite3/event_state_keys_table.go (hunk 2 above)
Comment (neilalexander) [reply roomserver/storage/sqlite3/event_state_keys_table.go:139]: Aren't they right now? `bulkSelectEventStateKeySQL` for `bulkSelectEventStateKey` (as it returns a state key), `bulkSelectEventStateKeyNIDSQL` in `bulkSelectEventStateKeyNID` (as it returns a NID)?

**Pull request merged**